File: frontend/src/sim/alu.js
export function simulateALU({ x, y, shiftDirection, AddSub, logicFunc, funcClass, constAmount, constVar }) {
        x &= 0b1111;
        y &= 0b1111;
    
        let shiftedY = 0b0000;
        let xorY = 0b0000;
        let xPlusY = 0b0000;
        let logic = 0b0000;
        let carry = 0;
        let s = 0b0000;
        let variableAmount = x & 0b11;
        let amount = 0b00;

        switch(constVar) {
            case 0b0:
                amount = constAmount;
            break;

            case 0b1:
                amount = variableAmount;//2 lsb of x
            break;
        }

        //shifter
        shiftedY = shiftDirection ? (y << amount) : (y >> amount);
        shiftedY &= 0b1111;

        //Adder
        const addSubMask = AddSub ? 0b1111 : 0b0000;
        xorY = y ^ addSubMask;

        const sum = x + xorY + AddSub;
        carry = (sum >> 4) & 1;
        xPlusY = sum & 0b1111;
        
        //Logic Unit
        switch (logicFunc) {
            case 0b00:
                logic = x & y;
            break;

            case 0b01:
                logic = x | y;
            break;

            case 0b10:
                logic = x ^ y;
            break;

            case 0b11:
                logic = ~(x | y) & 0b1111;
            break;
        }

        let msb = xPlusY > 0 ? 0b1111 : 0b0000;

        //Select Output
        switch (funcClass) {
            case 0b00:
                s = shiftedY;
            break;

            case 0b01:
                s = msb;
            break;

            case 0b10:
                s = xPlusY;
            break;

            case 0b11:
                s = logic;
            break;
        }

        s &= 0b1111;

        return {
            inputs: {
                x,
                y,
                shiftDirection,
                AddSub,
                logicFunc,
                funcClass,
                constAmount,
                constVar
            },

            internal: {
                shiftedY,
                xorY,
                xPlusY,
                logic,
                carry,
                variableAmount,
                amount,
                msb
            },

            output: {
            s
            }
        };

}

File: backend/src/ALU.cpp
#include "../include/ALU.hpp"

ALU::ALU(Wire* X, Wire* Y, Wire* AddSub, Wire* ConstAmount, Wire* ConstVar,
        Wire* ShiftFunc, Wire* LogicFunc, Wire* FuncClass,
            Wire& S, Wire& COUT)

    : x(X), y(Y), addSub(AddSub), constAmount(ConstAmount), constVar(ConstVar),
        shiftFunc(ShiftFunc), logicFunc(LogicFunc), funcClass(FuncClass),
        s(S), Cout(COUT),

        x5LSB(5), amount(5), yXOR(32), shiftedY(32), xPlusY(32), logic(32), xPlusYMsb(1),
        AddSub_SE(32), funcClass_s0(1), funcClass_s1(1),

        shiftAmount(constAmount, &x5LSB, constVar, amount),
        se(addSub, AddSub_SE),
        xorGate({y, &AddSub_SE}, yXOR),
        shifter(y, &amount, shiftFunc, shiftedY),
        adder(x, &yXOR, addSub, xPlusY, Cout),
        lu(x, y, logicFunc, logic),
        mux(&shiftedY, &xPlusYMsb, &xPlusY, &logic, &funcClass_s1, &funcClass_s0, s)
{}

void ALU::eval(){
    funcClass_s0.set(funcClass->getBit(0));
    funcClass_s1.set(funcClass->getBit(1));

    for (int i = 0; i < 5; i++) {
        x5LSB.setBit(i, x->getBit(i));
    }

    shiftAmount.eval();
    se.eval();
    xorGate.eval();
    shifter.eval();
    adder.eval();

    xPlusYMsb.set(xPlusY.getBit(31));
    
    lu.eval();
    mux.eval();
}

uint32_t ALU::getWireByPath(const std::string& path) {
    if (path.rfind("shiftAmount.", 0) == 0) {
        return shiftAmount.getWireByPath(path.substr(12));
    }
    if (path.rfind("adder.", 0) == 0) {
        return adder.getWireByPath(path.substr(6));
    }
    if (path.rfind("lu.", 0) == 0) {
        return lu.getWireByPath(path.substr(3));
    }
    if (path.rfind("mux.", 0) == 0) {
        return mux.getWireByPath(path.substr(4));
    }

    // INPUTS
    if (path == "x") return x->getValue();
    if (path == "y") return y->getValue();
    if (path == "addSub") return addSub->getValue();
    if (path == "constAmount") return constAmount->getValue();
    if (path == "constVar") return constVar->getValue();
    if (path == "shiftFunc") return shiftFunc->getValue();
    if (path == "logicFunc") return logicFunc->getValue();
    if (path == "funcClass") return funcClass->getValue();
    
    // INTERNAL
    if (path == "x5LSB") return x5LSB.getValue();
    if (path == "amount") return amount.getValue();
    if (path == "yXOR") return yXOR.getValue();
    if (path == "shiftedY") return shiftedY.getValue();
    if (path == "xPlusY") return xPlusY.getValue();
    if (path == "logic") return logic.getValue();
    if (path == "xPlusYMsb") return xPlusYMsb.getValue();
    
    // OUTPUTS
    if (path == "s") return s.getValue();
    if (path == "Cout") return Cout.getValue();

    return -1;
}